File: EMS Advanced Data Import 3.13 Full Source for Rad Studio 11 Alexandria/demos/CBuilder/DialogDemo/ImpDlgF.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ImpDlgF.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "QImport3Wizard"
#pragma resource "*.dfm"
TfmImpDlg *fmImpDlg;
//---------------------------------------------------------------------------
__fastcall TfmImpDlg::TfmImpDlg(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

String _DATA_DIR = "data";

void __fastcall TfmImpDlg::FormCreate(TObject *Sender)
{
  String dataDir = ExtractFileDir(Application->ExeName);
  int i = 0;
  while (!DirectoryExists(Format("%s\\%s", ARRAYOFCONST((dataDir, _DATA_DIR)))) && (i < 5))
  {
    dataDir = ExtractFileDir(dataDir);
    i++;
  }
  dataDir = Format("%s\\%s", ARRAYOFCONST((dataDir, _DATA_DIR)));

  Table1->LoadFromFile(Format("%s\\%s.xml", ARRAYOFCONST((dataDir, "country_empty"))));
}
//---------------------------------------------------------------------------
void __fastcall TfmImpDlg::FormDestroy(TObject *Sender)
{
  Table1->Active = False;
}
//---------------------------------------------------------------------------
void __fastcall TfmImpDlg::btImportClick(TObject *Sender)
{
  QImportWizard1->Execute();
}
//---------------------------------------------------------------------------
void __fastcall TfmImpDlg::QImportWizard1AfterImport(TObject *Sender)
{
  Table1->Close();
  Table1->Open();
}
//---------------------------------------------------------------------------
void __fastcall TfmImpDlg::QImportWizard1BeforePost(TObject *Sender,
  TQImportRow *Row, bool &Accept)
{
  if (chbUseBeforePost->Checked)
    for (int i = 0; i <= Row->Count - 1; i++)
      if (Row->Items[i]->Value == "Argentina")
      {
        Accept = false;
        return;
      }
}
//---------------------------------------------------------------------------
void __fastcall TfmImpDlg::Button1Click(TObject *Sender)
{
  if ((pcDestinations->ActivePage == tshDataSet) ||
      (pcDestinations->ActivePage == tshDBGrid))
  {
    Table1->DisableControls();
    try
    {
      Table1->First();
      while (!Table1->Eof)
      {
        Table1->Delete();
      }
    }
    __finally
    {
      Table1->EnableControls();
    }
  }
  else if (pcDestinations->ActivePage == tshListView)
  {
    ListView->Items->BeginUpdate();
    try
    {
      ListView->Items->Clear();
    }
    __finally
    {
      ListView->Items->EndUpdate();
    }
  }
  else if (pcDestinations->ActivePage == tshStringGrid)
  {
    for (int i = 1; i <= StringGrid->RowCount - 1; i++)
      for (int j = 0; j <= StringGrid->ColCount - 1; j++)
        StringGrid->Cells[j][i] = EmptyStr;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfmImpDlg::pcDestinationsChange(TObject *Sender)
{
  if (pcDestinations->ActivePage == tshDataSet)
    QImportWizard1->ImportDestination = qidDataSet;
  else if (pcDestinations->ActivePage == tshDBGrid)
    QImportWizard1->ImportDestination = qidDBGrid;
  else if (pcDestinations->ActivePage == tshListView)
    QImportWizard1->ImportDestination = qidListView;
  else if (pcDestinations->ActivePage == tshStringGrid) 
    QImportWizard1->ImportDestination = qidStringGrid;
}

File: EMS Advanced Data Import 3.13 Full Source for Rad Studio 11 Alexandria/demos/Delphi/DialogDemo/ImpDlgF.pas
unit ImpDlgF;

interface

uses
{$IF CompilerVersion >= 21}
  MidasLib,
{$IFEND}
  Forms, Db, Qimport3Wizard, Grids, DBGrids, StdCtrls, Classes,
  Controls, ExtCtrls, Qimport3, Qimport3Common, ComCtrls, Graphics, Windows, DBClient, SysUtils;

type
  TfmImpDlg = class(TForm)
    QImportWizard1: TQimport3Wizard;
    Panel1: TPanel;
    btImport: TButton;
    DataSource1: TDataSource;
    chbUseBeforePost: TCheckBox;
    Button1: TButton;
    pcDestinations: TPageControl;
    tshDataSet: TTabSheet;
    dgrDataSet: TDBGrid;
    tshDBGrid: TTabSheet;
    tshListView: TTabSheet;
    tshStringGrid: TTabSheet;
    DBGrid: TDBGrid;
    ListView: TListView;
    StringGrid: TStringGrid;
    Table1: TClientDataSet;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure btImportClick(Sender: TObject);
    procedure QImportWizard1AfterImport(Sender: TObject);
    procedure QImportWizard1BeforePost(Sender: TObject; Row: TQImportRow;
      var Accept: Boolean);
    procedure Button1Click(Sender: TObject);
    procedure pcDestinationsChange(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  fmImpDlg: TfmImpDlg;

implementation

{$R *.DFM}

const
  _DATA_DIR = 'data';

procedure TfmImpDlg.FormCreate(Sender: TObject);
var
  dataDir: string;
  i: Integer;
begin
  dataDir := ExtractFileDir(Application.ExeName);
  i := 0;
  while not DirectoryExists(Format('%s\%s', [dataDir, _DATA_DIR])) and (i < 5) do
  begin
    dataDir := ExtractFileDir(dataDir);
    Inc(i);
  end;
  dataDir := Format('%s\%s', [dataDir, _DATA_DIR]);
  Table1.LoadFromFile(Format('%s\%s.xml', [dataDir, 'country_empty']));

  StringGrid.Cells[0, 0] := 'Name of Country';
  StringGrid.Cells[1, 0] := 'Capital of Country';
  StringGrid.Cells[2, 0] := 'Continent of Country';
  StringGrid.Cells[3, 0] := 'Area of Country';
  StringGrid.Cells[4, 0] := 'Population of Country';
end;

procedure TfmImpDlg.FormDestroy(Sender: TObject);
begin
  Table1.Active := False;
end;

procedure TfmImpDlg.btImportClick(Sender: TObject);
begin
  QImportWizard1.Execute;  
end;

procedure TfmImpDlg.QImportWizard1AfterImport(Sender: TObject);
begin
  Table1.Close;
  Table1.Open;
end;

procedure TfmImpDlg.QImportWizard1BeforePost(Sender: TObject;
  Row: TQImportRow; var Accept: Boolean);
var
  i: integer;
begin
  if chbUseBeforePost.Checked then
    for i := 0 to Row.Count - 1 do
      if AnsiCompareText(Row[i].Value, 'Argentina') = 0 then begin
        Accept := false;
        Exit;
      end;
end;

procedure TfmImpDlg.Button1Click(Sender: TObject);
var
  i, j: integer;
begin
  if (pcDestinations.ActivePage = tshDataSet) or (pcDestinations.ActivePage = tshDBGrid) then
  begin
    Table1.DisableControls;
    try
      Table1.First;
      while not Table1.Eof do
        Table1.Delete;
    finally
      Table1.EnableControls;
    end;
  end
  else if pcDestinations.ActivePage = tshListView then
  begin
    ListView.Items.BeginUpdate;
    try
      ListView.Items.Clear;
    finally
      ListView.Items.EndUpdate;
    end;
  end
  else
    if pcDestinations.ActivePage = tshStringGrid then
    begin
      for i := 1 to StringGrid.RowCount - 1 do
        for j := 0 to StringGrid.ColCount - 1 do
          StringGrid.Cells[j, i] := EmptyStr;
      StringGrid.RowCount := 2;
    end;
end;

procedure TfmImpDlg.pcDestinationsChange(Sender: TObject);
begin
  if pcDestinations.ActivePage = tshDataSet then
    QImportWizard1.ImportDestination := qidDataSet
  else if pcDestinations.ActivePage = tshDBGrid then
    QImportWizard1.ImportDestination := qidDBGrid
  else if pcDestinations.ActivePage = tshListView then
    QImportWizard1.ImportDestination := qidListView
  else if pcDestinations.ActivePage = tshStringGrid then
    QImportWizard1.ImportDestination := qidStringGrid;
end;

end.
